### DISCUSSION TITLE
Rust interop demo

### DISCUSSION THREAD

**Discussion by alensiljak** [beancount/beancount#762] [General]
I've created a repository for experimenting with the Python-Rust interop: https://github.com/ledger-rs/rust-python-interop-demo
There's not much there right now. I'm basically playing around, trying to understand the various types in Pyo3 ([docs](https://pyo3.rs/v0.19.0/)). 
Below are links to some chapters that describe the interop in more detail:

- https://pyo3.rs/v0.19.0/types
- https://pyo3.rs/v0.19.0/conversions/tables

I was wondering if there is anything specific that would be interesting for Beancount to try out and demonstrate.
Based on the documentation, zero-copy is possible and in both memory spaces - Python and Rust. There seems to be many options available and I'm even getting lost trying to figure out what would be the best option to demonstrate.

Also, based on the previous discussion, I'm wondering if the time would be better used trying to work-out the real issues in Rust and see how to reuse the Python unit tests.

I'll continue playing around, to learn what is possible. Any feedback and ideas are welcome.

**Comment by alensiljak** [comment]
As per previous suggestion, I'll look at `beancount/beancount/core/data.proto` as the target structure.

**Comment by alensiljak** [comment]
Revisiting this topic. The link below shows the type equivalents in Python and Rust. No need for serialisation or other woodoo magic. All the Python values are references in Rust, hence a zero-cost interop. 

https://pyo3.rs/main/conversions/tables

>Using the Python-native types is almost zero-cost (they just require a type check similar to the Python builtin function isinstance()). 

I intend to dig into this more with a demo that reads SQLite in Python and uses that data in Rust.

**Comment by blais** [comment]
Sounds great, I'll definitely check it out
